viper.go

```diff
@@ -280,76 +279,67 @@ func WatchConfig() { v.WatchConfig() }
-			initWG.Done()
-			return
-		}
-
-		configFile := filepath.Clean(filename)
-		configDir, _ := filepath.Split(configFile)
-		realConfigFile, _ := filepath.EvalSymlinks(filename)
-
-		eventsWG := sync.WaitGroup{}
-		eventsWG.Add(1)
-		go func() {
-			for {
-				select {
-				case event, ok := <-watcher.Events:
-					if !ok { // 'Events' channel is closed
-						eventsWG.Done()
-						return
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok { // 'Events' channel is closed
+					return
+				}
+				currentConfigFile, _ := filepath.EvalSymlinks(filename)
+				// we only care about the config file with the following cases:
+				// 1 - if the config file was modified or created
+				// 2 - if the real path to the config file changed (eg: k8s ConfigMap replacement)
+				if (filepath.Clean(event.Name) == configFile &&
+					(event.Has(fsnotify.Write) || event.Has(fsnotify.Create))) ||
+					(currentConfigFile != "" && currentConfigFile != realConfigFile) {
+					realConfigFile = currentConfigFile
+					err := v.ReadInConfig()
+					if err != nil {
+						v.logger.Error(fmt.Sprintf("read config file: %s", err))
 					}
-					currentConfigFile, _ := filepath.EvalSymlinks(filename)
-					// we only care about the config file with the following cases:
-					// 1 - if the config file was modified or created
-					// 2 - if the real path to the config file changed (eg: k8s ConfigMap replacement)
-					if (filepath.Clean(event.Name) == configFile &&
-						(event.Has(fsnotify.Write) || event.Has(fsnotify.Create))) ||
-						(currentConfigFile != "" && currentConfigFile != realConfigFile) {
-						realConfigFile = currentConfigFile
-						err := v.ReadInConfig()
-						if err != nil {
-							v.logger.Error(fmt.Sprintf("read config file: %s", err))
-						}
-						if v.onConfigChange != nil {
-							v.onConfigChange(event)
-						}
-					} else if filepath.Clean(event.Name) == configFile && event.Has(fsnotify.Remove) {
-						eventsWG.Done()
-						return
+					if v.onConfigChange != nil {
+						v.onConfigChange(event)
 					}
-
-				case err, ok := <-watcher.Errors:
-					if ok { // 'Errors' channel is not closed
-						v.logger.Error(fmt.Sprintf("watcher error: %s", err))
-					}
-					eventsWG.Done()
+				} else if filepath.Clean(event.Name) == configFile && event.Has(fsnotify.Remove) {
 					return
 				}
+
+			case err, ok := <-watcher.Errors:
+				if ok { // 'Errors' channel is not closed
+					v.logger.Error(fmt.Sprintf("watcher error: %s", err))
+				}
+				return
 			}
-		}()
-		err = watcher.Add(configDir)
-		if err != nil {
-			v.logger.Error(fmt.Sprintf("failed to add watcher: %s", err))
-			initWG.Done()
-			return
 		}
-		initWG.Done()   // done initializing the watch in this go routine, so the parent routine can move on...
-		eventsWG.Wait() // now, wait for event loop to end in this go-routine...
 	}()
-	initWG.Wait() // make sure that the go routine above fully ended before returning
```

Comment:
I'm surprised by the changes about wait group here, and all the changes you did in the file to do not use waitgroup

I would have kept the change minimal to support removal the os.Exit, and that's it.

The other changes, which are good, I feel, are making the current PR uneasy to review.

I would have split this in 2 PRs one that can be easily be merged (the thing about panic), and a refactoring PR that could be reviewed and merged separately.

I'm not suggesting to change anything right now. I report here how I would have done it.
Other reviewers or you could have different opinions.

Also, I can be simply wrong and the changes here are somehow needed to address the os.Exit issue.

Reply:
Thanks for confirming what I thought about the fact the code had no need for such asynchronous thing.

You are right about the stack in the panic, it makes sense.

So here, I would like to suggest you to split the first commit in two.

- the first one to remove the complexity and keeping the os.Exit with a clear commit message about it's a refactoring made for simplifying over - complicated code
- then a commit about the change about os.Exit.

(They could be inverted of course)

This way the PR diff will stay unchanged, but at least the history would be way clearer